### PR TITLE
Support List/Show Functions (UDFs) and Describe Function <name>

### PR DESCRIPTION
**Pull request by dguy**
### Description 
Add new syntax `LIST FUNCTIONS;` , `SHOW FUNCTIONS;`, and `DESCRIBE FUNCTION <name>;`.
Will show a list of all UDFs (note doesn't include aggregations at the moment), i.e., 

```
 Function Name     
-------------------
 TIMESTAMPTOSTRING 
 TRIM              
 SUBSTRING         
 EXTRACTJSONFIELD  
 RANDOM            
 ROUND             
 CONCAT            
 ARRAYCONTAINS     
 UCASE             
 GEO_DISTANCE      
 FLOOR             
 ABS               
 LCASE             
 LEN               
 IFNULL            
 STRINGTOTIMESTAMP 
 CEIL              
-------------------
```

`DESCRIBE FUNCTION substring;`

```
Name        : SUBSTRING
Author      : Confluent
Version     : 
Overview    : returns a substring of the passed in value
Variations  : 

	Arguments   : VARCHAR, INT
	Returns     : VARCHAR
	Description : Returns a string that is a substring of this string. The substring begins with the character at the specified startIndex and extends to the end of this string.

	Arguments   : VARCHAR, INT, INT
	Returns     : VARCHAR
	Description : Returns a string that is a substring of this string. The substring begins with the character at the specified startIndex and extends to the character at endIndex -1.
```

### Testing done 
Added test to `CliTest` to ensure the results are parsed and printed.
Added test to `KsqlResourceTest`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by dguy** [comment]
Note: I just added description to the above output, though we may not want to include it in List/Show. Let me know what you think

**Comment by dguy** [comment]
Looking into the failure. Shouldn't stop review, however. Test works when i run in Idea, but not in build. So trying to work out what is going wrong

**Comment by blueedgenick** [comment]
not sure that just a big long list of all the available functions is super-helpful to a user ? maybe now that the descriptions are added ? **However**, i do see utility in being able to check which of my own custom udfs were successfully loaded and registered - how do you feel about either (a) changing this to only show the custom ones ? (those from the /ext location); or (b) having an optional `show external functions` variant to support this ?

**Comment by hjafarpour** [comment]
I was going to make similar suggestion to the one @blueedgenick 's. I think we can add a `DESCRIBE FUNCTION` command to show the details and maybe an example call for a given function. That way you can keep the list of functions concise and in case you need more details you can get details on individual function.

**Comment by apurvam** [comment]
I think `LIST FUNCTIONS` and `DESCRIBE FUNCTION FOO` is a good combination of commands.

the `DESCRIBE FUNCTION FOO` should print out the text of the UDF description annotation along with the arguments. This will make it much easier to discover UDFs and learn how to use them all within the KSQL CLI. 

Without `DESCRIBE FUNCTION` it will be really hard to present a usable description. I do think `LIST FUNCTIONS` should show _all_ the functions, maybe marking the external ones as such.

**Comment by dguy** [comment]
Yeah, i agree with @apurvam - i think we need both `LIST` and `DESCRIBE`. 
I also don't see why we should differentiate between external and internal. A function is a function is a function and it would be nice to be able to see what they are and what they do. We could mark them as external, but how does that help someone use the function? I don't think it adds any value.

**Comment by dguy** [comment]
Updated to add describe function. See the description above for the current format. Still need to add some tests for this

**Comment by blueedgenick** [comment]
@dguy i don't disagree that a function is a function, and providing help on what they are and how to sue them is great. I do think there are 2 different topics (trying to be) addressed here though: (1) help on what functions are available and their syntax; (2) UDFs in the sense of user-supplied code (are we settled on calling these "external" btw ? or something else - need a consistent terminology here) - consider the thought process of someone trying to build and load their own UDF, and how they want to "check that it's working". I wouldn't want to have to wade through all the help for built-in functionality trying to find my own organizations' custom additions, feels very clunky.

File: ksql-cli/src/test/java/io/confluent/ksql/CliTest.java [hunk 1]
```diff
@@ -530,4 +530,18 @@ public void shouldPrintErrorOnUnsupportedAPI() throws Exception {
...
+    final List<List<String>> rows = new ArrayList<>();
+    rows.add(Collections.singletonList("TIMESTAMPTOSTRING"));
+    rows.add(Collections.singletonList("EXTRACTJSONFIELD"));
+    rows.add(Collections.singletonList("CONCAT"));
+    testListOrShow("functions", TestResult.OrderedResult.build(rows), false);
+  }
+
+  @Test
+  public void shouldDescribeFunction() {
+    test("describe function timestamptostring;", EMPTY_RESULT);
```
Comment (apurvam) [inline ksql-cli/src/test/java/io/confluent/ksql/CliTest.java]: I assume this has to be filled in?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/DescribeFunctionList.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+package io.confluent.ksql.rest.entity;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+
+public class DescribeFunctionList extends KsqlEntity {
```
Comment (apurvam) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/DescribeFunctionList.java]: Would this be better named `FunctionDescriptionList`? Seems closer to what it actually is, and follows the `FunctionNameList` pattern.

File: ksql-cli/src/test/java/io/confluent/ksql/CliTest.java (hunk 1 above)
Comment (apurvam) [reply ksql-cli/src/test/java/io/confluent/ksql/CliTest.java]: We also need cases where the parameter to `DESCRIBE FUNCTION` references a non existent function.

Comment (dguy) [reply ksql-cli/src/test/java/io/confluent/ksql/CliTest.java]: yeah - i still have to finish of the tests...

**Comment by hjafarpour** [comment]
@dguy seems that after you moved the `SUBSTRING` function to external UDFs it's not available anymore. Can you move it back to the internal functions and use another simple function as en example for the UDF?

**Comment by dguy** [comment]
@hjafarpour - it is available.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/FunctionNameList.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+
+public class FunctionNameList extends KsqlEntity {
+
+  private final Collection<String> functionNames;
+
+  public FunctionNameList(
+      @JsonProperty("statementText") final String statementText,
```
Comment (hjafarpour) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/FunctionNameList.java]: `statementText` is not used.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java [hunk 4]
```diff
@@ -483,6 +499,35 @@ private KsqlEntity listTables(String statementText, boolean showDescriptions) {
...
+    return new FunctionNameList(statementText,
+        udfFactories
+            .stream()
+            .map(factory -> factory.getName().toUpperCase())
+            .collect(Collectors.toList())
+    );
+  }
+
+  private FunctionDescriptionList describeFunction(final String statementText,
+                                                   final String functionName) {
```
Comment (hjafarpour) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java]: nit:One arg per line

**Pull request merged**